controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -0,0 +1,312 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/blang/semver"
+	"github.com/coredns/corefile-migration/migration"
+	"github.com/docker/distribution/reference"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeadmv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/types/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	corefileKey       = "Corefile"
+	corefileBackupKey = "Corefile-backup"
+	coreDNSKey        = "coredns"
+)
+
+type coreDNSMigrator interface {
+	Migrate(currentVersion string, toVersion string, corefile string, deprecations bool) (string, error)
+}
+
+type CoreDNSMigrator struct{}
+
+func (c *CoreDNSMigrator) Migrate(fromCoreDNSVersion, toCoreDNSVersion, corefile string, deprecations bool) (string, error) {
+	return migration.Migrate(fromCoreDNSVersion, toCoreDNSVersion, corefile, deprecations)
+}
+
+type coreDNSInfo struct {
+	Corefile       string
+	Deployment     *appsv1.Deployment
+	CurrentVersion string
+}
+
+// UpdateCoreDNS updates the kubeadm configmap, coredns corefile and coredns
+// deployment.
+func (w *Workload) UpdateCoreDNS(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane) error {
+	info, err := w.getCoreDNSInfo(ctx)
+	if err != nil {
+		return err
+	}
+	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration == nil {
+		return nil
+	}
+
+	// Validate the image tag.
+	if err := validateCoreDNSImageTag(info.CurrentVersion, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag); err != nil {
+		return errors.Wrapf(err, "failed to validate CoreDNS")
+	}
+
+	// Get the DNS cluster configuration.
+	dns := &kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS
+
+	// Perform the upgrade.
+	if err := w.updateCoreDNSImageInfoInKubeadmConfigMap(ctx, dns); err != nil {
+		return err
+	}
+	if err := w.updateCoreDNSCorefile(ctx, info, dns); err != nil {
+		return err
+	}
+	if err := w.updateCoreDNSDeployment(ctx, info, dns); err != nil {
+		return errors.Wrap(err, "unable to update coredns deployment")
+	}
+	return nil
+}
+
+// validateCoreDNSImageTag returns error if the versions don't meet requirements.
+// Some of the checks come from
+// https://github.com/coredns/corefile-migration/blob/v1.0.6/migration/migrate.go#L414
+func validateCoreDNSImageTag(fromVersion, toVersion string) error {
```

Comment:
This check function can be omitted, since the `migration.Migrate` function already does the check before proceeding to handle the Corefile migration

Reply:
We need to move the check into the validation webhook for KCP

Reply:
Thanks! We might have to refactor the code later on. It was mentioned that we'll probably want to move the check to the validation webhook instead of having it here

Reply:
#2620